### ISSUE TITLE
Extend @internal syntax to encapsulate elements within a namespace

### ISSUE THREAD

**Issue by bdsl** [vimeo/psalm#1614]
It would be useful to be able to restrict the use of classes and other elements to within the library or submodule that they live in, particularly to allow changes to be made without breaking backwards compatibility. Although PHP has no concept of either a library or a submodule, the [@internal](https://docs.phpdoc.org/references/phpdoc/tags/internal.html) annotation can sometimes be used to achieve this.

Psalm interprets`@internal` as meaning that the declaring and calling code must not be in 'unrelated namepsaces', and emits issues as shown at https://psalm.dev/r/90a95c1dd8 . 

However this only works because the two namespaces are completely unrelated, and so might not be usable if `Foo` and `Bat` were both written by members of the same organisation, or in the particular case of Drupal if `Foo` was part of Drupal Core, and `Bat` was part of a bespoke applications controller, since the Drupal docs encourage the use of related namespaces.

So my proposal is to introduce a new syntax for internal, probably initially as `@psalm-internal`, which would take a a parameter to declare exactly what module the element is internal to. Psalm would act as shown below:

```php
<?php
namespace A\FooLibrary {
    /**
     * @psalm-internal A\FooLibrary
     */
    class Foo { }
}

namespace A\FooLibrary\Bar {
    class Bat {
        public function batBat(): void {
            new \A\FooLibrary\Foo(); // no issue emitted
        }
    }
}

namespace A {
    class Bat {
        public function batBat(): void {
            new \A\FooLibrary\Foo(); // proposed variant of InternalClass emitted
        }
    }
}

namespace A\FooLibraryEtc {
    class Bat {
        public function batBat(): void {
            new \A\FooLibrary\Foo(); // proposed variant of InternalClass emitted
        }
    }
}
```

I haven't contributed to Psalm before, but if there's interest in this I'd like to try to implement it and submit a PR.

**Comment by bdsl** [comment]
If I was to work on this I'd be tempted to also consolidate the 12 occurrences of the regex `'/^([^\\\]+).*/'`that are in the codebase as part of the implementation of `@internal`.

**Comment by bdsl** [comment]
So it needs more work, but I've got the start of an implementation. Currently it just stops a class extending a parent which is marked as internal to a different namespace. I'll implement checks for other types of usage and tidy stuff up a bit before I make the PR.

Would have been a bit quicker if I hadn't spent about half an hour debugging before I realised I'd put the docbloc on to the namespace instead of the class in the test code.

https://github.com/vimeo/psalm/compare/master...bdsl:psalm-internal?expand=0
